### DISCUSSION TITLE
Node.JS external authentication failure

### DISCUSSION THREAD

**Discussion by ghost** [processone/ejabberd#3573] [Questions] [answered]
I want to implement an external authentication in my ejabberd server. I am using 'ejabberd-auth' library for authentication. Technologies, database and OS used for this purpose:

Node.JS
Ejabberd as XMPP server
MySQL Database
React-Native (Front-End)
OS - Ubuntu 18.04

By far, I have achieved setting up ejabberd with ubuntu as user and registering the user and getting status from both the API and command line with base authentication. For example, I can register a user from any other device/system and from the command line too. But the problem with this is that all the data gets stored in ejabberd database on the Ubuntu server due to which there are data duplication chances also I already have a database structured and defined and already have data in that and so I choose external to go with external authentication.

Here is the confiiguration file of ejabberd.

> host_config:
> "experiment.org.co":
>     auth_method: [external]
>     extauth_program: "node /nodeauth/auth.js"
>     modules:
>       mod_register:
>         welcome_message:
>           subject: "Welcome!"
>           body: "Hi. Welcome to this XMPP server."
>       mod_stats: []
>       mod_time: []
>       mod_version: []
>   "localhost":
>     auth_method: [internal, anonymous]
> 
> listen:
>   -
>     port: 5280
>     ip: "::"
>     module: ejabberd_http
>     request_handlers:
>      "/api": mod_http_api
>     web_admin: true
> 
> acl:
>   admin:
>     user:
>       - "admin" : "experiment.org.co"
>   local:
>     user_regexp: ""
>   loopback:
>     ip:
>       - "127.0.0.0/8"
>       - "::1/128"
> 
> access_rules:
>   local:
>     - allow: local
>   c2s:
>     - deny: blocked
>     - allow
>   announce:
>     - allow: admin
>   configure:
>     - allow: admin
>   muc_create:
>     - allow: local
>   pubsub_createnode:
>     - allow: local
>   register:
>     - allow
>   trusted_network:
>     - allow: loopback
> 
> api_permissions:
>   "console commands":
>     from:
>       - ejabberd_ctl
>     who: all
>     what: "*"
>   "admin access":
>     who:
>       - access:
>           - allow:
>             - acl: loopback
>             - acl: admin
>     what:
>       - "*"
>       - "!stop"
>       - "!start"
>   "public commands":
>     who:
>       - ip: "127.0.0.1/8"
>     what:
>       - "status"
>       - "connected_users_number"

Here is the ejabberd.log when ejabberctl is started

>2019-03-26 13:36:05.844 [error] <0.427.0>@extauth:handle_info:149 Failed to start external authentication program 'node /nodeauth/auth.js'
> 2019-03-26 13:36:05.844 [error] <0.426.0> Supervisor 'extauth_pool_experimental.org.co' had child 'extauth_pool_experimental.org.co_1' started with extauth:start_link('extauth_pool_experimental.org.co_1', "node /nodeauth/auth.js") at <0.427.0> exit with reason normal in context child_terminated
> 2019-03-26 13:36:05.872 [warning] <0.360.0>@ejabberd_pkix:check_domain_certfiles:290 No certificate found matching 'experimental.org.co': strictly configured clients or servers will reject connections with this host; obtain a certificate for this (sub)domain from any trusted CA such as Let's Encrypt (www.letsencrypt.org)
> 2019-03-26 13:36:05.872 [warning] <0.360.0>@ejabberd_pkix:check_domain_certfiles:290 No certificate found matching 'localhost': strictly configured clients or servers will reject connections with this host; obtain a certificate for this (sub)domain from any trusted CA such as Let's Encrypt (www.letsencrypt.org)
> 2019-03-26 13:36:05.872 [info] <0.88.0>@ejabberd_app:start:60 ejabberd 19.02.28 is started in the node ejabberd@localhost in 1.48s
> 2019-03-26 13:36:05.874 [info] <0.33.0> Application ejabberd started on node ejabberd@localhost
> 2019-03-26 13:36:05.874 [info] <0.359.0>@ejabberd_listener:init:138 Start accepting TCP connections at [::]:5280 for ejabberd_http
> 2019-03-26 13:36:05.923 [error] <0.442.0>@extauth:handle_info:149 Failed to start external authentication program 'node /nodeauth/auth.js'
> 2019-03-26 13:36:05.923 [error] <0.426.0> Supervisor 'extauth_pool_experimental.org.co' had child 'extauth_pool_experimental.org.co_1' started with extauth:start_link('extauth_pool_experimental.org.co_1', "node /nodeauth/auth.js") at <0.442.0> exit with reason normal in context child_terminated
> 2019-03-26 13:36:05.924 [error] <0.426.0> Supervisor 'extauth_pool_experimental.org.co' had child 'extauth_pool_experimental.org.co_1' started with extauth:start_link('extauth_pool_experimental.org.co_1', "node /nodeauth/auth.js") at <0.442.0> exit with reason reached_max_restart_intensity in context shutdown

Here is error.log when ejabberd is started

> 2019-03-26 13:42:48.818 [error] <0.479.0>@extauth:handle_info:149 Failed to start external authentication program 'node /nodeauth/auth.js'
> 2019-03-26 13:42:48.818 [error] <0.477.0> Supervisor 'extauth_pool_experiment.org.co' had child 'extauth_pool_experiment.org.co_1' started with extauth:start_link('extauth_pool_experiment.org.co_1', "node /nodeauth/auth.js") at <0.479.0> exit with reason normal in context child_terminated
> 2019-03-26 13:42:48.818 [error] <0.477.0> Supervisor 'extauth_pool_experiment.org.co' had child 'extauth_pool_experiment.org.co_1' started with extauth:start_link('extauth_pool_experiment.org.co_1', "node /nodeauth/auth.js") at <0.479.0> exit with reason reached_max_restart_intensity in context shutdown
> 2019-03-26 13:42:48.818 [error] <0.392.0> Supervisor ejabberd_backend_sup had child 'extauth_pool_experiment.org.co' started with extauth_sup:start_link(<<"experiment.org.co">>, "node /nodeauth/auth.js", 'extauth_pool_experiment.org.co') at <0.477.0> exit with reason shutdown in context child_terminated
> 2019-03-26 13:42:48.818 [error] <0.392.0> Supervisor ejabberd_backend_sup had child 'extauth_pool_experiment.org.co' started with extauth_sup:start_link(<<"experiment.org.co">>, "node /nodeauth/auth.js", 'extauth_pool_experiment.org.co') at <0.477.0> exit with reason reached_max_restart_intensity in context shutdown
> 2019-03-26 13:42:48.819 [error] <0.338.0> Supervisor ejabberd_sup had child ejabberd_backend_sup started with ejabberd_backend_sup:start_link() at <0.392.0> exit with reason shutdown in context child_terminated

Any help regarding this topic will be appreciated. Please let me know if you need any more information regarding this.

**Comment by licaon-kter** [comment]
What ejabberd version?

You did see that `Failed to start external authentication program 'node /nodeauth/auth.js` in the log, right?

**Comment by ghost** [comment]
@licaon-kter I am using ejabberd 19.02.28. 

Yes, but I am not able to trace the reason behind that. I gave both relative and absolute path in the configuration file but still getting the error. All the permission are given also given to the file. Also if I run the file from the terminal. i.e. When I run the command node /nodeauth/auth.js the script is running. Also tried making this file a executable file and changed the following configuration in yml file
extauth_program: "./nodeauth/auth.js" but still getting the same error.

Is there anything I am missing other than this?

**Comment by licaon-kter** [comment]
That log is with `loglevel: 5` ?

What user do you use to run ejabberd?

What does `sudo -u userusedforejabberd stat /nodeauth/auth.js` output?

**Comment by ghost** [comment]
The log is with loglevel: 4.

I am using ubuntu user to run ejabberd.

Output for  sudo -u ubuntu stat /nodeauth/auth.js
> stat: cannot stat '/nodeauth/auth.js': No such file or directory

also 

Output for  sudo -u ubuntu stat /opt/ejabberd-server/etc/ejabberd/nodeauth/auth.js

> File: /opt/ejabberd-server/etc/ejabberd/nodeauth/auth.js
>   Size: 1502      	Blocks: 8          IO Block: 4096   regular file
> Device: ca01h/51713d	Inode: 1287752     Links: 1
> Access: (0777/-rwxrwxrwx)  Uid: ( 1000/  ubuntu)   Gid: ( 1000/  ubuntu)
> Access: 2019-03-26 18:27:17.372492869 +0000
> Modify: 2019-03-26 13:02:30.153508978 +0000
> Change: 2019-03-26 13:02:30.157508824 +0000
> Birth: -

Do you want me to provide new logs with loglevel: 5 and setting extauth_program: "node /opt/ejabberd-server/etc/ejabberd/nodeauth/auth.js" ?

**Comment by licaon-kter** [comment]
Well...you've guessed right. :)

**Comment by ghost** [comment, marked as answer]
@licaon-kter After changing the path to "node /opt/ejabberd-server/etc/ejabberd/nodeauth/auth.js" I am not able to replicate the above issue anymore. Thanks for the all the help you provided me. I am now able to successfully connect to my external script and call it when any command is executed through command line. 

But when I execute the API register call through Postman. I am using mod_http_api the external script is not being called. Below is my configuration for http

```
 listen:
   -
     port: 5280
     ip: "::"
     module: ejabberd_http
     request_handlers:
      "/api": mod_http_api
     web_admin: true

 api_permissions:
   "console commands":
     from:
       - ejabberd_ctl
     who: all
     what: "*"
   "admin access":
     who:
       - access:
           - allow:
             - acl: loopback
             - acl: admin
     what:
       - "*"
       - "!stop"
       - "!start"
   "public commands":
     who:
       - ip: "127.0.0.1/8"
     what:
       - "status"
       - "connected_users_number"
       - "register"

 modules:
   mod_http_api: {}
```

Is there any more configuration needed for api calls

**Comment by licaon-kter** [comment]
Maybe open another issue 

Logs?

**Comment by ghost** [comment]
**With API** ::

> 2019-03-29 09:12:37.812 [debug] <0.456.0>@ejabberd_http:init:151 S: [{[<<"api">>],mod_http_api},{[<<"admin">>],ejabberd_web_admin}]
> 2019-03-29 09:12:37.812 [debug] <0.456.0>@ejabberd_http:process_header:307 (#Port<0.13127>) http query: 'POST' <<"/api/register">>
> 2019-03-29 09:12:37.812 [debug] <0.456.0>@ejabberd_http:process:394 [<<"api">>,<<"register">>] matches [<<"api">>]
> 2019-03-29 09:12:37.812 [info] <0.361.0>@ejabberd_listener: accept:238 (<0.456.0>) Accepted connection ::ffff:122.170.68.221:54840 -> ::ffff:172.31.28.134:5280
> 2019-03-29 09:12:37.819 [info] <0.456.0>@mod_http_api:log:548 API call register [{<<"user">>,<<"test">>},{<<"host">>,<<"example.org">>},{<<"password">>,<<"test">>}] from ::ffff:122.170.68.221:54840
> 2019-03-29 09:12:37.819 [debug] <0.456.0>@mod_http_api:extract_auth:171 Invalid auth data: {error,invalid_auth}

**api data** ::
```
URL - http://url:5280/api/register
Headers
- Authorization : Basic authdata
- Content-Type: application/json
- X-Admin: true
Body
{
	"user": "test",
	"host": "example.org",
	"password": "test"
}
```

**With command(ejabberdctl)** ::

> 2019-03-29 09:09:47.992 [debug] <0.346.0>@ejabberd_access_permissions:handle_call:140 Command 'register' execution allowed by rule 'console commands' (CallerInfo=#{caller_module => ejabberd_ctl})
> 2019-03-29 09:09:47.992 [debug] <0.448.0>@ejabberd_commands:do_execute_command:494 Executing command ejabberd_admin:register with Args=[<<"abc">>,<<"example.org">>,<<"1231">>]
> 2019-03-29 09:10:17.994 [error] <0.448.0>@ejabberd_auth_external:failure:103 External authentication program failed when calling 'user_exists' for abc@example.org: timeout
> 2019-03-29 09:10:17.999 [error] <0.448.0>@ejabberd_auth_external:failure:103 External authentication program failed when calling 'try_register' for abc@example.org: disconnected
> 2019-03-29 09:10:18.008 [error] <0.428.0> Supervisor 'extauth_pool_example.org' had child 'extauth_pool_example.org_1' started with extauth:start_link('extauth_pool_example.ord_1', "node /opt/ejabberd-server/etc/ejabberd/nodeauth/auth.js") at <0.429.0> exit with reason normal in context child_terminated
> 2019-03-29 09:10:18.009 [debug] <0.428.0> Supervisor 'extauth_pool_example.org' started extauth:start_link('extauth_pool_example.org_1', "node /opt/ejabberd-server/etc/ejabberd/nodeauth/auth.js") at pid <0.449.0>

In this I already have tryregister function declared do I need to rename it to try_register ?? If you want I can also share my auth file
